packages/gatsby/src/schema/node-model.js

```diff
@@ -489,13 +490,21 @@ class LocalNodeModel {
           queryFields,
           actualFieldsToResolve
         )
-        if (!node.__gatsby_resolved) {
-          node.__gatsby_resolved = {}
-        }
-        resolvedNodes.set(
-          node.id,
-          _.merge(node.__gatsby_resolved, resolvedFields)
+        const previouslyResolvedFields =
+          previouslyResolvedNodes?.get(node.id) ?? {}
+
+        const mergedResolvedFields = _.merge(
+          previouslyResolvedFields,
+          resolvedFields
         )
+
+        resolvedNodes.set(node.id, mergedResolvedFields)
+
+        // `resolvedNodesCache` redux state is always source
+        // of truth. `node.__gatsby_resolved` might not persist
+        // (we mutate node loaded from LMDB) and is only optimization
+        // to avoid additional lookups in `resolvedNodesCache` WHEN it exists
+        node.__gatsby_resolved = mergedResolvedFields
```

Comment:
Important here is that `node.__gatsby_resolved` contains all materialization runs and not potentially just latest one. That's why I added `previouslyResolvedNodes` which reads from state and not from node (that ephemeral and can't be GCed at random moments)

We have checks in code that rely on this `__gatsby_resolved` being fully complete - for example https://github.com/gatsbyjs/gatsby/blob/e24ec76a9e564001b7669c96eae1b78a8f348cce/packages/gatsby/src/datastore/in-memory/indexing.ts#L362-L369